server/controllers/booksController.js

```diff
@@ -23,23 +34,158 @@ export default class BooksController {
    */
   static async addBook(req, res, next) {
     try {
-      const data = req.body;
       const {
         title,
         body,
         description,
         genre,
         pages,
         author,
-      } = data;
-      await bookModel.create(data);
+        hardcopy
+      } = req.body;
+      const bookGenre = await findOrCreateGenre(genre);
+      const bookAuthor = await findOrCreateAuthor(author);
+      const data = {
+        title,
+        body,
+        description,
+        pages,
+        hardcopy
+      };
+
+      data.genreId = bookGenre[0].id;
+      data.authorId = bookAuthor[0].id;
+      data.uploadedBy = req.user.id;
+      await create(data);
       return successfulRequest(res, 201, {
         bookTitle: title,
         bookBody: body,
         bookDescription: description,
-        bookGenre: genre,
+        bookGenre: bookGenre[0].name,
         bookPages: pages,
-        bookAuthor: author,
+        bookAuthor: bookAuthor[0].name
+      });
+    } catch (err) {
+      return next(err);
+    }
+  }
+
```

Comment:
You have done lot of house-keeping on the codebase; your efforts are commendable. Applause!

However, do you have any mechanism for preventing duplicate addition of books to the database? And if you have, how are you handling the error in such a way that user would know what went wrong?

Reply:
Thanks for raising that point, however, I raised this issue yesterday during the standups and we agreed that a library could have more than one copy of a book since it only creates the instance of the book in the DB but doesn't create a new author for it. So one of the books could be borrowed and the duplicate could still be available

Reply:
Okay... Thanks for your explanation. I guess I missed that point during the standup call